src/System.IO.FileSystem/src/System/IO/Win32FileSystem.cs

```diff
@@ -421,180 +421,150 @@ private static SafeFileHandle OpenHandle(string fullPath, bool asDirectory)
+                                    topLevel: false);
+                            }
+                            catch (Exception e)
+                            {
+                                if (exception == null)
+                                    exception = e;
+                            }
+                        }
+                        else
+                        {
+                            // Reparse point, don't recurse, just remove.
+                            if (findData.dwReserved0 == Interop.Kernel32.IOReparseOptions.IO_REPARSE_TAG_MOUNT_POINT)
+                            {
+                                // Mount point. Unmount using full path plus a trailing '\'.
+                                // (Note: This doesn't remove the underlying directory)
+                                string mountPoint = Path.Combine(fullPath, fileName + PathHelpers.DirectorySeparatorCharAsString);
+                                if (!Interop.Kernel32.DeleteVolumeMountPoint(mountPoint) && exception == null)
                                 {
                                     errorCode = Marshal.GetLastWin32Error();
-                                    if (errorCode != Interop.Errors.ERROR_PATH_NOT_FOUND)
+                                    if (errorCode != Interop.Errors.ERROR_SUCCESS && 
+                                        errorCode != Interop.Errors.ERROR_PATH_NOT_FOUND)
                                     {
-                                        try
-                                        {
-                                            throw Win32Marshal.GetExceptionForWin32Error(errorCode, data.cFileName);
-                                        }
-                                        catch (Exception e)
-                                        {
-                                            if (ex == null)
-                                                ex = e;
-                                        }
+                                        exception = Win32Marshal.GetExceptionForWin32Error(errorCode, fileName);
                                     }
                                 }
                             }
-                        }
-                        else
-                        {
-                            string fileName = Path.Combine(fullPath, data.cFileName);
-                            r = Interop.Kernel32.DeleteFile(fileName);
-                            if (!r)
+
+                            // Note that RemoveDirectory on a symbolic link will remove the link itself.
+                            if (!Interop.Kernel32.RemoveDirectory(Path.Combine(fullPath, fileName)) && exception == null)
                             {
                                 errorCode = Marshal.GetLastWin32Error();
-                                if (errorCode != Interop.Errors.ERROR_FILE_NOT_FOUND)
+                                if (errorCode != Interop.Errors.ERROR_PATH_NOT_FOUND)
```

Comment:
I could change it, but that would be a functional change. @ianhays what do you think?